telemetry/metrics.go

```diff
@@ -70,61 +33,87 @@ type GatherResponse struct {
 	ContentType string
 }
 
-// New creates a new instance of Metrics
-func New(cfg Config) (_ *Metrics, rerr error) {
-	if !cfg.Enabled {
+// New creates a new instance of metrics
+func New(opts ...Option) (Metrics, error) {
+	var err error
+
+	c := new(Config)
+	c.loadDefaults()
+	for _, fn := range opts {
+		err = fn(c)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if !c.Enabled {
 		return nil, nil
 	}
 
-	if numGlobalLables := len(cfg.GlobalLabels); numGlobalLables > 0 {
-		parsedGlobalLabels := make([]metrics.Label, numGlobalLables)
-		for i, gl := range cfg.GlobalLabels {
+	if numGlobalLables := len(c.GlobalLabels); numGlobalLables > 0 {
+		parsedGlobalLabels := make([]gometrics.Label, numGlobalLables)
+		for i, gl := range c.GlobalLabels {
 			parsedGlobalLabels[i] = NewLabel(gl[0], gl[1])
 		}
 
-		globalLabels = parsedGlobalLabels
+		c.globalLabels = parsedGlobalLabels
 	}
 
-	metricsConf := metrics.DefaultConfig(cfg.ServiceName)
-	metricsConf.EnableHostname = cfg.EnableHostname
-	metricsConf.EnableHostnameLabel = cfg.EnableHostnameLabel
+	metricsConf := gometrics.DefaultConfig(c.ServiceName)
+	metricsConf.EnableHostname = c.EnableHostname
+	metricsConf.EnableHostnameLabel = c.EnableHostnameLabel
 
-	memSink := metrics.NewInmemSink(10*time.Second, time.Minute)
-	inMemSig := metrics.DefaultInmemSignal(memSink)
-	defer func() {
-		if rerr != nil {
-			inMemSig.Stop()
-		}
-	}()
+	memSink := gometrics.NewInmemSink(10*time.Second, time.Minute)
+	gometrics.DefaultInmemSignal(memSink)
 
-	m := &Metrics{memSink: memSink}
-	fanout := metrics.FanoutSink{memSink}
+	m := &metrics{memSink: memSink, cnf: c}
+	fanout := gometrics.FanoutSink{memSink}
+	var promSink *metricsprom.PrometheusSink
 
-	if cfg.PrometheusRetentionTime > 0 {
-		m.prometheusEnabled = true
-		prometheusOpts := metricsprom.PrometheusOpts{
-			Expiration: time.Duration(cfg.PrometheusRetentionTime) * time.Second,
-		}
+	//Initialize the external library only once.
+	initOnce.Do(func() {
```

Comment:
mhhh this doesn't seem right to me. Calling `New` probably shouldn't have such `Once` behavior.

You create a local `*metrics` object and then call `Once` on it. What happens when I call `New` again? This seems like a bad design of a constructor. Instead, I would consider moving the `Once` call to the `Init` function, it makes more sense there IMO.

Reply:
this is because of Singleton issues in the external package, for example, duplication registration on `/metrics`.
let me see what can I do.

Reply:
resolved